ambry-store/src/main/java/com/github/ambry/store/BootstrapController.java

```diff
@@ -0,0 +1,203 @@
+        // 1. this is the first time to add new replica onto current node;
+        // 2. last replica addition failed at some point before updating InstanceConfig in Helix
+        if (isFileCopyFeatureEnabled()) {
+          // "New partition -> FC"
+          listenerToInvoke = fileCopyManagerListener;
+          logStateChange("New partition -> FC", partitionName);
+        } else {
+          // "New partition -> R"
+          listenerToInvoke = storageManagerListener;
+          logStateChange("New partition -> R", partitionName);
+        }
+      } else {
+        if (isFileCopyFeatureEnabled()) {
+          if (isFileExists(replica.getPartitionId(), BOOTSTRAP_IN_PROGRESS_FILE_NAME)) {
+            // R.Incomplete -> FC
+            listenerToInvoke = storageManagerListener;
+            logStateChange("R.Incomplete -> FC", partitionName);
+          } else if (isAnyLogSegmentExists(replica.getPartitionId())) {
+            if (isFileExists(replica.getPartitionId(), FILECOPY_IN_PROGRESS_FILE_NAME)) {
+              // FC.Incomplete -> FC
+              listenerToInvoke = fileCopyManagerListener;
+              logStateChange("FC.Incomplete -> FC", partitionName);
+            } else {
+              // R.complete -> FC or FC.complete -> FC
+              listenerToInvoke = storageManagerListener;
+              logStateChange("R.complete -> FC or FC.complete -> FC", partitionName);
+            }
+          }
+        } else {
+          if (isFileExists(replica.getPartitionId(), BOOTSTRAP_IN_PROGRESS_FILE_NAME)) {
+            // R.Incomplete -> R
+            listenerToInvoke = storageManagerListener;
+            logStateChange("R.Incomplete -> R", partitionName);
+          } else if (isAnyLogSegmentExists(replica.getPartitionId())) {
+            if (isFileExists(replica.getPartitionId(), FILECOPY_IN_PROGRESS_FILE_NAME)) {
+              // FC.Incomplete -> R
+              try {
+                deleteFileCopyData(replica.getPartitionId());
+              } catch (IOException | StoreException e) {
+                String message = "Failed `deleteFileCopyData` step for " + partitionName;
+                logger.error(message);
+                throw new StateTransitionException(message, BootstrapControllerFailure);
+              }
+              listenerToInvoke = storageManagerListener;
+              logStateChange("FC.Incomplete -> R", partitionName);
+            } else {
+              // R.complete -> R or FC.complete -> R
+              listenerToInvoke = storageManagerListener;
+              logStateChange("R.complete -> R or FC.complete -> R", partitionName);
+            }
+          }
+        }
+      }
+
+      assert listenerToInvoke != null;
+      listenerToInvoke.onPartitionBecomeBootstrapFromOffline(partitionName);
+      if (listenerToInvoke == fileCopyManagerListener) {
+        try {
+          storeManager.getPrimaryClusterParticipant().getReplicaSyncUpManager().waitForFileCopyCompleted(partitionName);
```

Comment:
Can we pass replica synup manager object to this class directly?

Reply:
+1

Reply:
`replicaSyncUpManager` is a prop of `HelixParticipant` class. 
And we need to involve `replicaSyncUpManager. waitForFileCopyCompleted()` for the storeManager's ClusterParticipant. 
How can we pass  `waitForFileCopyCompleted` object to BootstrapController class?